### PR TITLE
Add extra logging around GitHub API calls

### PR DESCRIPTION
**Pull request by garath**
Resolves dotnet/arcade#8579.

Maestro occasionally errors because of responses from GitHub. The current logging situation isn't giving quite enough info to determine if this is a Maestro problem or a GitHub problem. Application Insights telemetry shows that Octokit is throwing `ApiValidationException`s, but the default message and stack is missing context.

This PR adds extra logging around calls to capture the information GitHub sends in its response JSON. This should give much more detailed information on what exactly went wrong. Hopefully then that will be enough to know how to guard against them.

Info on the error payload format may be found at [Resources in the REST API - Client errors](https://docs.github.com/en/rest/overview/resources-in-the-rest-api#client-errors).

### CODE REVIEW

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 1]
```diff
@@ -203,7 +203,7 @@ public async Task ProcessDependencyUpdateErrorsAsync()
                                 break;
                             }
 
-                            _logger.LogInformation($"Matching issue {issueNumber} for {repoBranchUpdateHistoryError.Repository}:{repoBranchUpdateHistoryError.Branch} was closed. Creating a new issue.");
+                            _logger.LogInformation($"Matching issue {issueNumber.Value} for {repoBranchUpdateHistoryError.Repository}:{repoBranchUpdateHistoryError.Branch} was closed. Creating a new issue.");
```
Comment (ChadNedzlek) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:206]: Really these should use the "structured logging" thing, and not interpolated strings, then the "issueNumber" would be available as a detail of the log.
```suggestion
                            _logger.LogInformation("Matching issue {IssueNumber} for {Repository}:{Branch} was closed. Creating a new issue.", issueNumber, repoBranchUpdateHistoryError.Repository, repoBranchUpdateHistoryError.Branch);
```
Nullables should ToString just fine, I think.

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs [hunk 2]
```diff
@@ -422,14 +424,14 @@ public async Task MergeDependencyPullRequestAsync(string pullRequestUrl, MergePu
...
+            IssueComment lastComment = (await ExecuteAndLogApiErrors(() => Client.Issue.Comment.GetAllForIssue(owner, repo, id))).LastOrDefault();
             if (lastComment != null && lastComment.Body.EndsWith(CommentMarker))
             {
-                await Client.Issue.Comment.Update(owner, repo, lastComment.Id, message + CommentMarker);
+                await ExecuteAndLogApiErrors(() => Client.Issue.Comment.Update(owner, repo, lastComment.Id, message + CommentMarker));
             }
             else
             {
-                await Client.Issue.Comment.Create(owner, repo, id, message + CommentMarker);
+                await ExecuteAndLogApiErrors(() => Client.Issue.Comment.Create(owner, repo, id, message + CommentMarker));
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:434]: Why make all these calls catch and log? Why not just log at the higher level?  Catching/logging should always be done at the "highest" level possible, so that you don't have to wrap every single call to every method like this.

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs [hunk 3]
```diff
@@ -1283,5 +1290,65 @@ public async Task DeletePullRequestBranchAsync(string pullRequestUri)
...
+        /// the standard exception message. See <see href="https://docs.github.com/en/rest/overview/resources-in-the-rest-api#client-errors">Client Errors</see> 
+        /// for details on the structure.
+        /// </summary>
+        private void LogApiErrorDetails(ApiException apiException)
+        {
+            _logger.LogError(apiException, "Caught ApiException, will rethrow -- StatusCode: {StatusCode}", apiException.StatusCode);
+
+            if (!string.IsNullOrEmpty(apiException.ApiError.DocumentationUrl))
+            {
+                _logger.LogError("Documentation URL: {DocumentationUrl}", apiException.ApiError.DocumentationUrl);
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:1307]: These should probably be just be details of the original log. It can be hard to correlate 3 error logs together, since they can get interspersed with many others.

**Review by ChadNedzlek — CHANGES_REQUESTED** [review]

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs (hunk 1 above)
Comment (garath) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:206]: In this form it doesn't stringify properly, that's why I added this. Though I guess I could add a good `.ToString()` to the class.

Comment (garath) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:206]: Interpolated logging strings are present all over these classes, so I wanted to just limit the scope of changes. I didn't see much value in including _all_ interpolation-to-structured logging changes in this PR, but will if you think it's worth it.

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs (hunk 3 above)
Comment (garath) [reply src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:1307]: This method also logs all "error" entries from the response, so even more lines. 

https://github.com/dotnet/arcade-services/blob/6e4f7ff49dbe81a3b568ab56c96da4249b2fa9ba/src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs#L1310-L1315

My thought was that the operation IDs are enough to isolate instances. 

Open to other ideas, of course. I felt that just using a single log message would be harder to make sense of.

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs (hunk 2 above)
Comment (garath) [reply src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:434]: I had a nice grep :-)

But yeah, for these methods that make more than one call with the GitHub client it's probably nicer to just have one try-catch.

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs (hunk 3 above)
Comment (mmitche) [reply src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:1307]: I seem to remember a single operation id didn't correlate all that tightly...

If we expect that there will be not too many error details typically? < 2 perhaps? then I agree with @ChadNedzlek  that including these in a single message will be better.

Comment (garath) [reply src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:1307]: Fair. 

How would you feel if I serialized it all to JSON? It's weird because it's the only message like that. It's good because Kusto could make sense of it.

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs (hunk 1 above)
Comment (ChadNedzlek) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:206]: At least for ones you're touching, updating them is cheap and can't hurt.

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs (hunk 2 above)
Comment (ChadNedzlek) [reply src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:434]: Even that's probably not high enough.  I'd put it WAAAAY out at the entry point, and only have a single try/catch anywhere.

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs (hunk 3 above)
Comment (ChadNedzlek) [reply src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs]: Just add them as structured logging values.  Then you can get them with "customDimention.DocumentationUrl" and such.  Unfortunately, it would be best to just have them as properties on the trace, but ILogger doesn't expose that. Basically ALL these logs are just adding additional properties, the message itself isn't really all that useful, so I'd just have the message be `"StatusCode: {StatusCode}, DocumentaionUrl: {DocumentationUrl}, Code: {Code}"` and such.

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs (hunk 1 above)
Comment (garath) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs]: Went ahead and covered this whole class.

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs (hunk 3 above)
Comment (garath) [reply src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs]: I'm struggling a bit with this feedback. Making a single log message makes sense, no problem there. But the entirety of the information I'm trying to capture is too complex for an ad-hoc format string. To me it just makes a lot of sense to borrow JSON as a way to structure the data. And as an extra win, it will handle however few or many items actually end up in the details list. The JSON itself will appear in the structured logging fields, so super easy to parse out of the log. The only risk that I see is that throwing the serializer in at this point in the program crashing could be considered expensive, but... it's not really? 

Happy to consider more here but I'll need some help understanding alternatives.

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs (hunk 2 above)
Comment (garath) [reply src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs]: Fine point. I originally didn't like that a higher-level class would now need to know about a low-level (and implementation-specific) Octokit class, but the distance really isn't much at all and the wins in simplicity are clear. So I moved this up to the top level.

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 4]
```diff
@@ -242,10 +255,10 @@ public async Task ProcessDependencyUpdateErrorsAsync()
                                 break;
                             }
 
-                            _logger.LogInformation($"Matching issue {issueNumber} for subscription {subscriptionUpdateHistoryError.SubscriptionId} was closed. Creating a new issue.");
+                            _logger.LogInformation("Matching issue {IssueNumber} for subscription {SubscriptionId} was closed. Creating a new issue.", issueNumber.Value, subscriptionUpdateHistoryError.SubscriptionId);
```
Comment (mmitche) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:258]: does String() need to be called here? SubscriptionId is a Guid.

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs (hunk 3 above)
Comment (ChadNedzlek) [reply src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs]: The "DocumentationUrl" can just go in the first message. Let it be null/empty if it's null/empty.  The others, being a list per exception, probably isn't feasible to put inline, having them as a JSON blob works well enough. :-)

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs (hunk 2 above)
Comment (ChadNedzlek) [reply src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs]: Yeah, it's an... interesting problem.  C# doesn't have a good pattern of "I don't want to leak the details of this exception out to my callers" without try/catch everywhere.

If you wanted to get REALLY fancy, you could twist this whole things on it's head, and rather than changing what you log, instead inject something into the Logging/AppInsights pipeline to log this sort of thing special (sort of like we do for the HttpClient in all our AzDO calls).

Basically make it so that every exception of this type always gets logged fancy.  You could use an ITelemetryInitializer to handle this, and it could just get injected into the AI pipeline at the same place other GitHub stuff is.

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs (hunk 4 above)
Comment (ChadNedzlek) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:258]: Much like the interpolated thing, it's going to ToString, everything. Though GUID's have lots of formats, so it probably makes sense to call ToString yourself to get the format you want, and not let it just be whatever C# feels like.

Comment (mmitche) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:258]: Fine by me then. Just making sure we didn't get `System.Guid`

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs (hunk 3 above)
Comment (garath) [reply src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs]: Ah, okay, I see where we are now. Yeah, that list is actually the most interesting piece of info to capture. The other fields are good but the list is where the action is.

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs (hunk 4 above)
Comment (garath) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:258]: Its default ToString uses the "D" specifier, which gives a pleasant form like `382c74c3-721d-4f34-80e5-57657b6cbc27`

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs (hunk 3 above)
Comment (garath) [reply src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs]: I reformatted it a bit. Let me know if you approve (or better, let GitHub know)

**Review by ChadNedzlek — APPROVED** [review]
